File: src/iac_code/a2a/exposure.py
from __future__ import annotations

from collections.abc import Iterable
from enum import Enum
from typing import Any


class A2AExposureType(str, Enum):
    RAW_THINKING = "raw_thinking"
    TOOL_TRACE = "tool_trace"


_EXPOSURE_TYPE_ORDER = (
    A2AExposureType.RAW_THINKING,
    A2AExposureType.TOOL_TRACE,
)
_SUPPORTED_TYPE_NAMES = ", ".join(item.value.replace("_", "-") for item in _EXPOSURE_TYPE_ORDER)
_DISABLED_ALIASES = frozenset({"", "none", "off", "false", "0"})
_ALL_ALIASES = frozenset({"all", "*"})
DEFAULT_A2A_EXPOSURE_TYPES = frozenset({A2AExposureType.TOOL_TRACE})


def normalize_a2a_exposure_types(value: Any) -> frozenset[A2AExposureType]:
    if value is None:
        return DEFAULT_A2A_EXPOSURE_TYPES

    tokens = list(_iter_exposure_tokens(value))
    if not tokens:
        return frozenset()
    if any(token in _ALL_ALIASES for token in tokens):
        return frozenset(_EXPOSURE_TYPE_ORDER)
    tokens = [token for token in tokens if token not in _DISABLED_ALIASES]
    return frozenset(_exposure_type_from_token(token) for token in tokens)


def format_a2a_exposure_types(value: Any) -> list[str]:
    enabled = normalize_a2a_exposure_types(value)
    return [item.value for item in _EXPOSURE_TYPE_ORDER if item in enabled]


def _iter_exposure_tokens(value: Any) -> Iterable[str]:
    if isinstance(value, A2AExposureType):
        yield value.value
        return
    if isinstance(value, str):
        for item in value.replace(";", ",").split(","):
            token = item.strip().lower().replace("-", "_")
            if token:
                yield token
        return
    if isinstance(value, Iterable) and not isinstance(value, dict):
        for item in value:
            yield from _iter_exposure_tokens(item)
        return
    raise ValueError(
        f"A2A thinking exposure must be a string or list of strings. Supported values: {_SUPPORTED_TYPE_NAMES}."
    )


def _exposure_type_from_token(token: str) -> A2AExposureType:
    try:
        return A2AExposureType(token)
    except ValueError as exc:
        raise ValueError(
            f"Unsupported A2A thinking exposure type {token!r}. Supported values: {_SUPPORTED_TYPE_NAMES}."
        ) from exc

File: iac-code-rs/crates/iac-code-a2a/src/exposure.rs
use std::collections::BTreeSet;
use std::fmt;

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum A2AExposureType {
    RawThinking,
    ToolTrace,
}

impl A2AExposureType {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::RawThinking => "raw_thinking",
            Self::ToolTrace => "tool_trace",
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct A2AExposureError {
    message: String,
}

impl A2AExposureError {
    fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

impl fmt::Display for A2AExposureError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.message)
    }
}

impl std::error::Error for A2AExposureError {}

pub fn normalize_a2a_exposure_types(
    value: Option<&str>,
) -> Result<BTreeSet<A2AExposureType>, A2AExposureError> {
    match value {
        None => Ok(BTreeSet::from([A2AExposureType::ToolTrace])),
        Some(value) => normalize_a2a_exposure_tokens(split_tokens(value)),
    }
}

pub fn normalize_a2a_exposure_tokens<I, S>(
    values: I,
) -> Result<BTreeSet<A2AExposureType>, A2AExposureError>
where
    I: IntoIterator<Item = S>,
    S: AsRef<str>,
{
    let tokens = values
        .into_iter()
        .flat_map(|value| split_tokens(value.as_ref()))
        .collect::<Vec<_>>();
    if tokens.is_empty() {
        return Ok(BTreeSet::new());
    }
    if tokens
        .iter()
        .any(|token| matches!(token.as_str(), "all" | "*"))
    {
        return Ok(exposure_type_order().into_iter().collect());
    }

    let mut result = BTreeSet::new();
    for token in tokens {
        if matches!(token.as_str(), "" | "none" | "off" | "false" | "0") {
            continue;
        }
        result.insert(exposure_type_from_token(&token)?);
    }
    Ok(result)
}

pub fn format_a2a_exposure_types(values: &BTreeSet<A2AExposureType>) -> Vec<&'static str> {
    exposure_type_order()
        .into_iter()
        .filter(|value| values.contains(value))
        .map(A2AExposureType::as_str)
        .collect()
}

pub fn format_a2a_exposure_slice(values: &[A2AExposureType]) -> Vec<&'static str> {
    let values = values.iter().copied().collect::<BTreeSet<_>>();
    format_a2a_exposure_types(&values)
}

fn split_tokens(value: &str) -> Vec<String> {
    value
        .replace(';', ",")
        .split(',')
        .map(|item| item.trim().to_ascii_lowercase().replace('-', "_"))
        .filter(|item| !item.is_empty())
        .collect()
}

fn exposure_type_from_token(token: &str) -> Result<A2AExposureType, A2AExposureError> {
    match token {
        "raw_thinking" => Ok(A2AExposureType::RawThinking),
        "tool_trace" => Ok(A2AExposureType::ToolTrace),
        _ => Err(A2AExposureError::new(format!(
            "Unsupported A2A thinking exposure type {token:?}. Supported values: raw-thinking, tool-trace."
        ))),
    }
}

fn exposure_type_order() -> [A2AExposureType; 2] {
    [A2AExposureType::RawThinking, A2AExposureType::ToolTrace]
}
